snuba/snuba_migrations/spans/0001_spans_v1.py

```diff
@@ -0,0 +1,99 @@
+from typing import List, Sequence
+
+from snuba.clickhouse.columns import UUID, Column, DateTime, Nested, String, UInt
+from snuba.clusters.storage_sets import StorageSetKey
+from snuba.migrations import migration, operations, table_engines
+from snuba.migrations.columns import MigrationModifiers as Modifiers
+from snuba.migrations.operations import OperationTarget, SqlOperation
+from snuba.utils.schemas import Float
+
+storage_set_name = StorageSetKey.SPANS
+local_table_name = "spans_local"
+dist_table_name = "spans_dist"
+
+UNKNOWN_SPAN_STATUS = 2
+
+columns: List[Column[Modifiers]] = [
+    Column("project_id", UInt(64)),
+    Column("transaction_id", UUID(Modifiers(nullable=True))),
+    Column("trace_id", UUID()),
+    Column("span_id", UInt(64)),
+    Column("parent_span_id", UInt(64, Modifiers(nullable=True))),
+    Column("segment_id", UInt(64)),
+    Column("is_segment", UInt(8)),
+    Column("segment_name", String(Modifiers(default="''"))),
+    Column("start_timestamp", DateTime()),
+    Column("end_timestamp", DateTime()),
+    Column("timestamp", DateTime(Modifiers(materialized="end_timestamp"))),
+    Column("duration", UInt(32)),
+    Column("exclusive_time", Float(64)),
+    Column("op", String(Modifiers(low_cardinality=True))),
+    Column("group", UInt(64)),
+    Column("span_status", UInt(8)),
+    Column("span_kind", String(Modifiers(low_cardinality=True))),
+    Column("description", String()),
+    Column("status", UInt(32, Modifiers(nullable=True))),
+    Column("module", String(Modifiers(low_cardinality=True, nullable=True))),
+    Column("action", String(Modifiers(low_cardinality=True, nullable=True))),
+    Column("domain", String(Modifiers(nullable=True))),
+    Column("platform", String(Modifiers(low_cardinality=True, nullable=True))),
+    Column("user", String(Modifiers(nullable=True))),
+    Column("tags", Nested([("key", String()), ("value", String())])),
+    Column(
+        "measurements",
+        Nested(
+            [("key", String(Modifiers(low_cardinality=True))), ("value", Float(64))]
+        ),
+    ),
+    Column("partition", UInt(16)),
+    Column("offset", UInt(64)),
+    Column("retention_days", UInt(16)),
+    Column("deleted", UInt(8)),
+]
+
+
+class Migration(migration.ClickhouseNodeMigration):
+    blocking = False
+
+    def forwards_ops(self) -> Sequence[SqlOperation]:
+        return [
+            operations.CreateTable(
+                storage_set=storage_set_name,
+                table_name=local_table_name,
+                columns=columns,
+                engine=table_engines.ReplacingMergeTree(
+                    order_by="(project_id, segment_name, segment_id, timestamp)",
```

Comment:
I'm not sure if it's a good idea to order by `segment_name` if it defaults to `''`. Also, would it make sense to have `trace_id` in this?

Reply:
Regarding `trace_id`, I would be curious to know between which columns we would put it in the existing order by key. 

One of the reasons I did not add `trace_id` is because the primary use case I considered is finding segments of interest. Traces span across different segments. With the current sorting key, parts of a trace would be in random `segment_names`. We could potentially add `trace_id` after `segment_id` and also add some sort of skip index so that finding all spans of a trace would be faster and the skip index could filter out lots of granules because `trace_id` is sorted. But I am not sure if there is any advantage to doing that compared with just adding a skip index on the column. The timestamp field in the sorting key would already be efficient in filtering out granules which would not have the `trace_id` since traces are short lived.

Reply:
Ah, OK that all makes sense. If the primary thing is segment_id then leaving trace_id out makes sense to me too.